src/validation.cpp

```diff
@@ -2669,8 +2669,11 @@ static bool ReceivedBlockTransactions(const CBlock &block, CValidationState& sta
                 LOCK(cs_nBlockSequenceId);
                 pindex->nSequenceId = nBlockSequenceId++;
             }
-            if (chainActive.Tip() == nullptr || !setBlockIndexCandidates.value_comp()(pindex, chainActive.Tip())) {
-                setBlockIndexCandidates.insert(pindex);
+            {
+                LOCK(cs_main);
```

Comment:
I think we can just add an `AssertLockHeld(cs_main)` to the top of this function -- this function only has two callers, both of which lock cs_main before calling.